yabt/builders/cpp.py

```diff
@@ -21,38 +21,169 @@
 ~~~~~~~~~~~~~~~~
 
 :author: Itamar Ostricher
+
+TODO: support injecting conpile/link flags for 3rd party libs
+TODO: support flavors
 """
 
 
+from os.path import basename, dirname, join, relpath, splitext
+
 from .dockerapp import build_app_docker_and_bin, register_app_builder_sig
 from ..extend import (
-    PropType as PT, register_build_func, register_manipulate_target_hook)
+    PropType as PT, register_build_func, register_builder_sig,
+    register_manipulate_target_hook)
 from ..logging import make_logger
-from ..utils import yprint
+from ..target_utils import split
+from ..utils import link_artifacts, yprint
 
 
 logger = make_logger(__name__)
 
 
-register_app_builder_sig('CppApp', [('executable', PT.File)])
+register_app_builder_sig(
+    'CppApp', [('executable', PT.File, None), ('main', PT.Target, None)])
 
 
 @register_manipulate_target_hook('CppApp')
 def cpp_app_manipulate_target(build_context, target):
     logger.debug('Injecting "{}" to deps of {}',
                  target.props.base_image, target)
     target.deps.append(target.props.base_image)
+    if target.props.main and target.props.main not in target.deps:
+        logger.debug('Injecting "{}" to deps of {}',
+                     target.props.main, target)
+        target.deps.append(target.props.main)
 
 
 @register_build_func('CppApp')
-def cpp_bin_builder(build_context, target):
+def cpp_app_builder(build_context, target):
     """Pack a C++ binary as a Docker image with its runtime dependencies.
 
     TODO(itamar): Dynamically analyze the binary and copy shared objects
     from its buildenv image to the runtime image, unless they're installed.
     """
     yprint(build_context.conf, 'Build CppApp', target)
-    if target.props.executable not in target.artifacts['app']:
-        target.artifacts['app'].append(target.props.executable)
+    if target.props.executable and target.props.main:
+        raise KeyError(
+            '`main` and `executable` arguments are mutually exclusive')
+    if target.props.executable:
```

Comment:
I don't understand the difference between main and executable

Reply:
Backward compatibility. Currently we're using 'executable' to specify a 'named file path' to run (supporting binaries from external sources, like scons). The intent is to switch to 'main' that takes a target name instead of file path, for binaries built by ybt, so ybt takes care of tracking files from targets.

Reply:
gotcha